tests/functional/test_download.py

```diff
@@ -0,0 +1,315 @@
+        self.tempdir = tempfile.mkdtemp()
+        self.filename = os.path.join(self.tempdir, 'myfile')
+
+        # Initialize some default arguments
+        self.bucket = 'mybucket'
+        self.key = 'mykey'
+        self.extra_args = {}
+        self.subscribers = []
+
+        # Create a stream to read from
+        self.content = b'my content'
+        self.stream = six.BytesIO(self.content)
+
+    def tearDown(self):
+        super(BaseDownloadTest, self).tearDown()
+        shutil.rmtree(self.tempdir)
+
+    @property
+    def manager(self):
+        return self._manager
+
+    @property
+    def method(self):
+        return self.manager.download
+
+    @property
+    def call_kwargs(self):
+        return {
+            'bucket': self.bucket,
+            'key': self.key,
+            'fileobj': self.filename
+        }
+
+    @property
+    def invalid_extra_args(self):
+        return {
+            'Foo': 'bar'
+        }
+
+    @property
+    def stubbed_responses(self):
+        return [
+            {
+                'method': 'head_object',
+                'service_response': {
+                    'ContentLength': len(self.content)
+                }
+            },
+            {
+                'method': 'get_object',
+                'service_response': {
+                    'Body': self.stream
+                }
+            }
+        ]
+
+    @property
+    def expected_progress_callback_info(self):
+        # Note that last read is from the empty sentinel indicating
+        # that the stream is done.
+        return [
+            {'bytes_transferred': 10},
+            {'bytes_transferred': 0}
+        ]
+
+    def test_download_temporary_file_does_not_exist(self):
+        for stubbed_response in self.stubbed_responses:
+            self.stubber.add_response(**stubbed_response)
+
+        future = self.manager.download(**self.call_kwargs)
+        future.result()
+        # Make sure the file exists
+        self.assertTrue(os.path.exists(self.filename))
+        # Make sure the random temporary file does not exist
+        possible_matches = glob.glob('%s*' % self.filename + os.extsep)
+        self.assertEqual(possible_matches, [])
+
+    def test_download_cleanup_on_failure(self):
+        # Add the stubbed HeadObject response
+        self.stubber.add_response(**self.stubbed_responses[0])
+
+        # Throw an error on the download
+        self.stubber.add_client_error('get_object')
+
+        future = self.manager.download(**self.call_kwargs)
+
+        with self.assertRaises(ClientError):
+            future.result()
+        # Make sure the actual file and the temporary do not exist
+        # by globbing for the file and any of its extensions
+        possible_matches = glob.glob('%s*' % self.filename)
+        self.assertEqual(possible_matches, [])
+
+    def test_download_with_nonexistent_directory(self):
+        for stubbed_response in self.stubbed_responses:
```

Comment:
Can we put a more specific name here?  Ideally someone can just read the test body to understand what the test is doing:
```
self.respond_with_single_download() # or whatever you want to call it
...
future = self.manager.download(...)
```